crates/holochain_integrity_types/src/header.rs

```diff
@@ -111,13 +106,51 @@ macro_rules! write_into_header {
     };
 }
 
-/// A trait to specify the common parts of a Header
-pub trait HeaderInner {
-    /// Get a full header from the subset
+/// A trait to unify the "inner" parts of a Header, i.e. the structs inside
+/// the Header enum's variants. This trait is used for the "weighed" version
+/// of each struct, i.e. the version without weight information erased.
+///
+/// Header types with no weight are considered "weighed" and "unweighed" at the
+/// same time, but types with weight have distinct types for the weighed and
+/// unweighed versions.
+pub trait HeaderWeighed {
+    type Unweighed: HeaderUnweighed;
+    type Weight: Default;
+
+    /// Construct the full Header enum with this variant.
     fn into_header(self) -> Header;
+
+    /// Erase the rate limiting weight info, creating an "unweighed" version
+    /// of this header. This is used primarily by validators who need to run the
+    /// `weigh` callback on a header they received, and want to make sure their
+    /// callback is not using the predefined weight to influence the result.
+    fn unweighed(self) -> Self::Unweighed;
```

Comment:
Just out of curiosity... why would a validator ever need to run the `weigh` callback if they didn't want the weight to be a factor?

Reply:
Part of element validation now includes verifying that the weight on the header is correct, so the element authority has to weigh the op they've received, and to do that deterministically, they have to weigh the same header that the author weighed, which had no weight.

So, they have to read what the declared weight is, then erase that, weight it again, and see if the two match as part of validation.